### PR TITLE
Code Challenges

### PR DESCRIPTION
**Pull request by Moh-Awadallah**

### CODE REVIEW

File: codeChallenges/codeChallenge1/index.test.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+const arrOfNum = [1, 2, 2, 4, 5, 6, 6];
+const getUniqueValues = (arrOfNum) => {
+    let arr = [];  // create an empty array
+    for (let i=0, l=arrOfNum.length; i<l; i++) // iterate through
+        if (arr.indexOf(arrOfNum[i]) === -1) // using indexOf we check if the number is in the new array
+            arr.push(arrOfNum[i]); // if not then we push it
+    return arr; // reutn the array
+};
+console.log(getUniqueValues(arrOfNum)); // [1, 2, 4, 5, 6]
+console.log(getUniqueValues([1 , 2 , 1 , 4]));
```
Comment (jadabuawwad) [inline codeChallenges/codeChallenge1/index.test.js]: It seems here you are duplicated the console log, 

so try to final check and tidy your code.

**Review by jadabuawwad — COMMENTED** [review]
Challenge 1 review : 

 - Postive points : 
  - You are writing a comments which seems you have a tidy mindset.
  - You Got the solution.
  
  - Negative Points : 
    - You didn't check your final code consoling.
    - The title of branch have to be semantic as it could for example : "Code Challenge - 1 : Get unique values of array elements" also try to follow conventional commit from the resource provided.
    
    Final Mark is : 4/5

**Review by jadabuawwad — APPROVED** [review]
Challenge 2 review :

Postive points :
- You are following the optimal solution, Great!

Final Mark is : 5/5

**Review by jadabuawwad — COMMENTED** [review]
Challenge 3 review :

Postive points :
You Got the solution.

Negative Points :
You came late for tow days ^_^

Final Mark is : 5/5

**Review by jadabuawwad — CHANGES_REQUESTED** [review]
Challenge 4 review :

Postive points :
You are dealing with concepts of regex, string methods.

Negative Points : 
You Got the solution.

Negative Points :
You did'nt got a full solution except still there some commented code also.

Final Mark is : 3.7/5

File: codeChallenges/codeChallenge4/index.test.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * The first word within the output should be capitalized only if the original word was capitalized (known as Upper Camel Case, also often referred to as Pascal case).
+ */
+/**
+ * Examples
+  "theStealthWarrior"  gets converted to "the-stealth-warrior"
+  "TheStealthWarrior"  gets converted to "The_Stealth_Warrior"
+ */
+function toSnakeCase(str) {
+    
+ let newStr = str.replace(/[A-Z]/g, (char) => '_'+`${char.toLowerCase()}`)
```
Comment (jadabuawwad) [inline codeChallenges/codeChallenge4/index.test.js:12]: You are doing great here !

File: codeChallenges/codeChallenge4/index.test.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ return newStr.charAt(0) + newStr.substring(1);
```
Comment (jadabuawwad) [inline codeChallenges/codeChallenge4/index.test.js:13]: here in case if the code detect the first character as capital it will convert it into lower case with _ adding.

so the result is a string starts with _ , this is an edge case should be resolved.

File: codeChallenges/codeChallenge4/index.test.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ 
+//  //    let arr = str.split('')
```
Comment (jadabuawwad) [inline codeChallenges/codeChallenge4/index.test.js:15]: clean these stuff

File: codeChallenges/codeChallenge5/index.test.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    for (var i = 0; i < steps; i++) {
+      array.unshift(array.pop());
+    }
+    return array;
+  }
+  if (steps < 0) {
+    steps = Math.abs(steps);
+   return [...array.slice(steps), ...array.slice(0, steps)];
+  }
+
```
Comment (jadabuawwad) [inline codeChallenges/codeChallenge5/index.test.js:38]: What about the case when step is zero ?

**Review by jadabuawwad — COMMENTED** [review]
Challenge 5 review :

Postive points :
You got the concept great!

Negative Points :
On case is not covered when the step is zero.

**Review by jadabuawwad — APPROVED** [review]
Challenge 6 review :

Postive points :
You got the concept great!

Challenge 7 review :

Postive points :
You got the optimal solution which find tow patterns in the right way.

**Review by jadabuawwad — COMMENTED** [review]
Challenge 8 review :

Postive points :
The shortest solution but not the efficient

Negative points :
the complexity is to high

**Review by jadabuawwad — COMMENTED** [review]
Challenge 9 review :

Positive points:
Solution works Great !

Challenge 10 review :

Positive points:
Solution works Great !

**Review by jadabuawwad — APPROVED** [review]
Challenge 11 review :

Positive points:
You Got the One Line Solution which is GREAT!

**Review by jadabuawwad — APPROVED** [review]
Challenge 12 review :

Positive points:
You Got The solution but you need to improve readability by using array method like foreach

**Review by jadabuawwad — APPROVED** [review]
Code Challenge 14 Review:

You got the solution , is there a way to declare and assign multiple variables is JS ?
